### PR TITLE
238 bug attributeerror kwavesimulation object has no attribute record u split field

### LINKED ISSUE #238: [BUG] AttributeError: 'kWaveSimulation' object has no attribute 'record_u_split_field'
hello
when i wanna to imply my matlab simulation to the kwave-python, i get the following error:
```
WARNING:root:Highest prime factors in each dimension are [23 23]
WARNING:root:Use dimension sizes with lower prime factors to improve speed
Traceback (most recent call last):
  File "D:\HISLab\毕设\CODE\DATA\mydataset.py", line 76, in <module>
    sensor_data = kspace_first_order_2d_gpu(
  File "D:\HISLab\MODEL\Github\k-wave-python-master\kwave\kspaceFirstOrder2D.py", line 64, in kspace_first_order_2d_gpu
    sensor_data = kspaceFirstOrder2DC(
  File "D:\HISLab\MODEL\Github\k-wave-python-master\kwave\kspaceFirstOrder2D.py", line 132, in kspaceFirstOrder2DC
    sensor_data = kspaceFirstOrder2D(
  File "D:\HISLab\MODEL\Github\k-wave-python-master\kwave\kspaceFirstOrder2D.py", line 305, in kspaceFirstOrder2D
    k_sim.input_checking('kspaceFirstOrder2D')
  File "D:\HISLab\MODEL\Github\k-wave-python-master\kwave\kWaveSimulation.py", line 515, in input_checking
    self.create_sensor_variables()
  File "D:\HISLab\MODEL\Github\k-wave-python-master\kwave\kWaveSimulation.py", line 1381, in create_sensor_variables
    'record_u_split_field': self.record_u_split_field,
AttributeError: 'kWaveSimulation' object has no attribute 'record_u_split_field'
```

this is my simulation code in python,  i do the same step like kwave in matlab
```
import mat73
import scipy
import numpy as np
from kwave.data import Vector
from kwave.kgrid import kWaveGrid
from kwave.kmedium import kWaveMedium
from kwave.ksensor import kSensor
from kwave.ksource import kSource
from kwave.kspaceFirstOrder2D import kspace_first_order_2d_gpu
from kwave.options.simulation_execution_options import SimulationExecutionOptions
from kwave.options.simulation_options import SimulationOptions
from kwave.utils.conversion import cart2grid
from kwave.utils.kwave_array import kWaveArray
from kwave.utils.mapgen import make_cart_circle, make_disc
from kwave.utils.signals import reorder_binary_sensor_data

phantom_path = '../phantom.mat'
density_map_path = '../density_map.mat'
phantom = mat73.loadmat(phantom_path)['phantom']
density_map = mat73.loadmat(density_map_path)['density_map']



# grid
PMLSize = Vector([20, 20])
grid_dsize = Vector([1e-4, 1e-4])
grid_pixel = Vector([512, 512])
kgrid = kWaveGrid(grid_pixel, grid_dsize)

# medium
medium = kWaveMedium(
    sound_speed=1.5*density_map,
    density=density_map,
    alpha_coeff=0.75,
    alpha_power=1.5,
    BonA=6
)
kgrid.makeTime(medium.sound_speed)

# sensor
sensor = kSensor()
sensor_mask = np.zeros(grid_pixel)
sensor_radius = 200
sensor_angle = 2*np.pi
sensor_pos = grid_dsize / 2
sensor_points = 100
sensor_mask_idx = make_cart_circle(
    radius=sensor_radius,
    num_points=sensor_points,
    arc_angle=sensor_angle,
).astype(np.int16)
for x, y in sensor_mask_idx.T:
    sensor_mask[x, y] = 1
sensor.mask = sensor_mask

# source
source = kSource()
x = 250
y = 300
source_radius = 5
disc = make_disc(
    grid_size=grid_pixel,
    center=Vector([x, y]),
    radius=source_radius
)

source.p0 = phantom + disc

# simulation
simulation_options = SimulationOptions(
    pml_size=PMLSize,
    pml_inside=False,
)
execution_options = SimulationExecutionOptions(is_gpu_simulation=True)

sensor_data = kspace_first_order_2d_gpu(
    kgrid=kgrid,
    source=source,
    sensor=sensor,
    medium=medium,
    simulation_options=simulation_options,
    execution_options=execution_options)
```
thanks

### PR DESCRIPTION
**Pull request by waltsims**
Here, I try to add a more explicit error for users who do not set save_to_disk to true.

It could make sense to also try the following:
1. Set default value to True
2. remove save_to_disk as a user setting

This PR also fixes a bug that _always_ runs 2D simulations on the CPU.

closes #238

### CODE REVIEW

**Review by faridyagubbayli — APPROVED** [review]
LGTM! 

If we want to take one extra step, we can convert assertions to `raise (...)` statements. It is recommended to use assertions for internal code logic only. User inputs should be checked with `if-else` statements. [This article](https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/) explains a possible problem when using assertions to validate user inputs. There was another article that discussed this topic from code structure PoV but I can't find it.

**Comment by waltsims** [comment]
> LGTM!
> 
> If we want to take one extra step, we can convert assertions to `raise (...)` statements. It is recommended to use assertions for internal code logic only. User inputs should be checked with `if-else` statements. [This article](https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/) explains a possible problem when using assertions to validate user inputs. There was another article that discussed this topic from code structure PoV but I can't find it.

Thanks for bringing this to my attention. With this new knowledge, this PR now looks sloppy. I'll replace the assertions with raise statements to clean this up.

File: kwave/executor.py [hunk 1]
```diff
@@ -29,13 +29,13 @@ def run_simulation(self, input_filename: str, output_filename: str, options: str
...
         stdout = None if self.execution_options.show_sim_log else subprocess.DEVNULL
-        return_code = subprocess.run(command, stdout=stdout, shell=True).returncode
-
         try:
-            assert return_code == 0, f'Simulation call returned code: {return_code}'
-        except AssertionError:
-            if isinstance(return_code, unittest.mock.MagicMock):
+            subprocess.run(command, stdout=stdout, shell=True, check=True)
+        except subprocess.CalledProcessError as e:
+            if isinstance(e.returncode, unittest.mock.MagicMock):
```
Comment (faridyagubbayli) [inline kwave/executor.py:35]: This looks awesome!

File: kwave/kspaceFirstOrder2D.py [hunk 2]
```diff
@@ -295,13 +295,21 @@ def kspaceFirstOrder2D(
     # start the timer and store the start time
     TicToc.tic()
 
+    # Currently we only support binary execution, meaning all simulations must be saved to disk.
```
Comment (faridyagubbayli) [inline kwave/kspaceFirstOrder2D.py:298]: An idea that is not a blocker for this PR: what if we add this to the error string or somewhere in the documentation? Explanation here deserves more than staying as a comment IMO.

File: kwave/kspaceFirstOrder2D.py [hunk 3]
```diff
@@ -295,13 +295,21 @@ def kspaceFirstOrder2D(
...
+    if not simulation_options.save_to_disk:
+        if execution_options.is_gpu_simulation:
+            raise ValueError('GPU simulation requires saving to disk. Please set SimulationOptions.save_to_disk=True')
+        else:
+            raise ValueError('CPU simulation requires saving to disk. Please set SimulationOptions.save_to_disk=True')
```
Comment (faridyagubbayli) [inline kwave/kspaceFirstOrder2D.py:303]:
```suggestion
        target_accelerator = "GPU" if execution_options.is_gpu_simulation else "CPU"
        raise ValueError(f'{target_accelerator} simulation requires saving to disk. Please set SimulationOptions.save_to_disk=True')
```

File: kwave/executor.py (hunk 1 above)
Comment (waltsims) [reply kwave/executor.py:35]: That is some review 😅

File: kwave/kspaceFirstOrder3D.py [hunk 4]
```diff
@@ -298,6 +298,13 @@ def kspaceFirstOrder3D(
     # start the timer and store the start time
     TicToc.tic()
 
+    # Currently we only support binary execution, meaning all simulations must be saved to disk.
+    if not simulation_options.save_to_disk:
+        if execution_options.is_gpu_simulation:
+            raise ValueError('GPU simulation requires saving to disk. Please set SimulationOptions.save_to_disk=True')
+        else:
+            raise ValueError('CPU simulation requires saving to disk. Please set SimulationOptions.save_to_disk=True')
+
```
Comment (faridyagubbayli) [inline kwave/kspaceFirstOrder3D.py:307]: Repeated code, perhaps move to a utility function?

**Review by faridyagubbayli — APPROVED** [review]
Left some comments that are not blocking

File: kwave/kspaceFirstOrder2D.py (hunk 2 above)
Comment (waltsims) [reply kwave/kspaceFirstOrder2D.py:298]: I agree. We should open an enhancement issue for the next point release.

File: kwave/kspaceFirstOrder2D.py (hunk 3 above)
Comment (waltsims) [reply kwave/kspaceFirstOrder2D.py:303]: Would have been clean but my more explicit formulation was also ok I feel.

File: kwave/kspaceFirstOrder3D.py (hunk 4 above)
Comment (waltsims) [reply kwave/kspaceFirstOrder3D.py:307]: I already have an issue open to refactor the 2D3D functions due to a lot of repeated code. Next point release?

**Pull request merged**